community/io/src/main/java/org/neo4j/io/pagecache/impl/SingleFilePageSwapperFactory.java

```diff
@@ -78,6 +84,37 @@ public void renameUnopenedFile( File sourceFile, File targetFile, CopyOption...
     }
 
     @Override
+    public Stream<FileHandle> streamFilesRecursive( File directory ) throws IOException
+    {
+        try
+        {
+            // We grab a snapshot of the file tree to avoid seeing the same file twice or more due to renames.
+            List<File> snapshot = streamFilesRecursiveInner( directory ).collect( toList() );
```

Comment:
Yes, we need to snapshot because file renames could otherwise make files appear more than once in the stream.